File: src/com/intellij/idea/plugin/hybris/impex/formatting/ImpexBlock.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.impex.formatting;

import com.intellij.formatting.*;
import com.intellij.idea.plugin.hybris.impex.psi.ImpexTypes;
import com.intellij.lang.ASTNode;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.psi.TokenType;
import com.intellij.psi.codeStyle.CodeStyleSettings;
import com.intellij.psi.formatter.common.AbstractBlock;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Created 22:21 21 December 2014
 *
 * @author Alexander Bartash <AlexanderBartash@gmail.com>
 */
public class ImpexBlock extends AbstractBlock {

    private final SpacingBuilder spacingBuilder;
    private final CodeStyleSettings codeStyleSettings;
    private final AlignmentStrategy alignmentStrategy;

    public ImpexBlock(
        @NotNull final ASTNode node,
        @Nullable final Wrap wrap,
        @Nullable final Alignment alignment,
        @NotNull final SpacingBuilder spacingBuilder,
        @NotNull final CodeStyleSettings codeStyleSettings,
        @NotNull final AlignmentStrategy alignmentStrategy) {
        super(node, wrap, alignment);

        this.spacingBuilder = spacingBuilder;
        this.codeStyleSettings = codeStyleSettings;
        this.alignmentStrategy = alignmentStrategy;
    }

    @Override
    protected List<Block> buildChildren() {
        final List<Block> blocks = new ArrayList<Block>();

        alignmentStrategy.processNode(myNode);

        ASTNode currentNode = myNode.getFirstChildNode();

        while (null != currentNode) {
            alignmentStrategy.processNode(currentNode);

            if (isNotWhitespaceOrNewLine(currentNode) && !isCurrentNodeHasParentValue(currentNode)) {
                final var block = new ImpexBlock(
                    currentNode,
                    null,
                    alignmentStrategy.getAlignment(currentNode),
                    spacingBuilder,
                    codeStyleSettings,
                    alignmentStrategy
                );

                blocks.add(block);
            }

            currentNode = currentNode.getTreeNext();
        }

        return blocks;
    }

    private boolean isNotWhitespaceOrNewLine(final ASTNode currentNode) {
        return TokenType.WHITE_SPACE != currentNode.getElementType()
            && ImpexTypes.CRLF != currentNode.getElementType();
    }

    private boolean isCurrentNodeHasParentValue(final @NotNull ASTNode currentNode) {
        return Objects.equals(currentNode.getTreeParent().getElementType(), ImpexTypes.VALUE);
    }

    @Override
    public Indent getIndent() {
        return Indent.getNoneIndent();
    }

    @Nullable
    @Override
    public Spacing getSpacing(@Nullable final Block child1, @NotNull final Block child2) {
        return spacingBuilder.getSpacing(this, child1, child2);
    }

    @Override
    public boolean isLeaf() {
        return myNode.getFirstChildNode() == null;
    }
}

File: src/com/intellij/idea/plugin/hybris/impex/formatting/ImpexBlock.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for IntelliJ IDEA.
 * Copyright (C) 2019-2025 EPAM Systems <hybrisideaplugin@epam.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.impex.formatting


import com.intellij.formatting.*
import com.intellij.idea.plugin.hybris.impex.psi.ImpexTypes
import com.intellij.lang.ASTNode
import com.intellij.psi.TokenType
import com.intellij.psi.codeStyle.CodeStyleSettings
import com.intellij.psi.formatter.common.AbstractBlock

class ImpexBlock(
    node: ASTNode,
    wrap: Wrap? = Wrap.createWrap(WrapType.NONE, false),
    alignment: Alignment? = null,
    private val spacingBuilder: SpacingBuilder,
    private val codeStyleSettings: CodeStyleSettings,
    private val alignmentStrategy: ImpExAlignmentStrategy
) : AbstractBlock(node, wrap, alignment) {

    override fun getDebugName() = when (myNode.elementType) {
        ImpexTypes.HEADER_LINE -> "Header Line"
        ImpexTypes.VALUE_LINE -> "Value Line"
        ImpexTypes.ANY_HEADER_MODE -> "Mode"
        ImpexTypes.FULL_HEADER_PARAMETER -> "Parameter"
        ImpexTypes.ANY_HEADER_PARAMETER_NAME -> "Parameter Name"
        ImpexTypes.FULL_HEADER_TYPE -> "Type"
        ImpexTypes.HEADER_TYPE_NAME -> "Name"
        ImpexTypes.SUB_TYPE_NAME -> "Sub-Type"
        ImpexTypes.PARAMETERS_SEPARATOR -> "Separator"
        ImpexTypes.FIELD_VALUE_SEPARATOR -> "Separator"
        ImpexTypes.VALUE_GROUP -> "Value Group"
        ImpexTypes.VALUE -> "Value"
        ImpexTypes.LINE_COMMENT -> "Line Comment"
        ImpexTypes.MODIFIERS -> "Modifiers"
        ImpexTypes.ANY_ATTRIBUTE_NAME -> "Name"
        ImpexTypes.ANY_ATTRIBUTE_VALUE -> "Value"
        ImpexTypes.ASSIGN_VALUE -> "="
        ImpexTypes.ATTRIBUTE -> "Attribute"
        ImpexTypes.ATTRIBUTE_SEPARATOR -> ","
        ImpexTypes.LEFT_SQUARE_BRACKET -> "["
        ImpexTypes.RIGHT_SQUARE_BRACKET -> "]"
        ImpexTypes.MACRO_USAGE_DEC -> "Macro Usage"
        else -> "Block"
    }

    override fun isLeaf() = myNode.firstChildNode == null
    override fun getSpacing(child1: Block?, child2: Block): Spacing? = spacingBuilder.getSpacing(this, child1, child2)
    override fun getIndent(): Indent = Indent.getNoneIndent()

    override fun buildChildren(): List<Block> {
        val blocks = mutableListOf<Block>()

        alignmentStrategy.processNode(myNode)

        var currentNode = myNode.firstChildNode

        while (currentNode != null) {
            alignmentStrategy.processNode(currentNode)

            if (isNewBlockToBeMade(currentNode)) {
                val block = ImpexBlock(
                    node = currentNode,
                    alignment = alignmentStrategy.getAlignment(currentNode),
                    spacingBuilder = spacingBuilder,
                    codeStyleSettings = codeStyleSettings,
                    alignmentStrategy = alignmentStrategy
                )
                blocks.add(block)
            }

            currentNode = currentNode.treeNext
        }

        return blocks
    }

    private fun isNewBlockToBeMade(currentNode: ASTNode) = currentNode.elementType != TokenType.WHITE_SPACE
        && currentNode.elementType != ImpexTypes.CRLF
        && currentNode.elementType != ImpexTypes.ATTRIBUTE_NAME
        && currentNode.elementType != ImpexTypes.ATTRIBUTE_VALUE
        && currentNode.elementType != ImpexTypes.HEADER_TYPE
        && currentNode.elementType != ImpexTypes.VALUE_SUBTYPE
        && currentNode.treeParent.elementType != ImpexTypes.VALUE
        && currentNode.treeParent.elementType != ImpexTypes.ANY_ATTRIBUTE_VALUE

}